crates/lsp/src/types.rs

```diff
@@ -0,0 +1,187 @@
+    }
+
+    #[derive(Debug)]
+    pub(crate) struct NargoTests;
+    impl Request for NargoTests {
+        type Params = NargoTestsParams;
+        type Result = NargoTestsResult;
+        const METHOD: &'static str = "nargo/tests";
+    }
+}
+
+pub(crate) mod notification {
+    use lsp_types::notification::Notification;
+
+    use super::NargoPackageTests;
+
+    // Re-providing lsp_types that we don't need to override
+    pub(crate) use lsp_types::notification::{
+        DidChangeConfiguration, DidChangeTextDocument, DidCloseTextDocument, DidOpenTextDocument,
+        DidSaveTextDocument, Exit, Initialized,
+    };
+
+    pub(crate) struct NargoUpdateTests;
+    impl Notification for NargoUpdateTests {
+        type Params = NargoPackageTests;
+        const METHOD: &'static str = "nargo/tests/update";
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct NargoTestsOptions {
+    /// Tests can be requested from the server.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) fetch: Option<bool>,
+
+    /// Tests runs can be requested from the server.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) run: Option<bool>,
+
+    /// The server will send notifications to update tests.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) update: Option<bool>,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct NargoCapability {
+    /// The server will provide various features related to testing within Nargo.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) tests: Option<NargoTestsOptions>,
+}
+
+#[derive(Debug, PartialEq, Clone, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct ServerCapabilities {
+    /// Defines how text documents are synced.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) text_document_sync: Option<TextDocumentSyncCapability>,
+
+    /// The server provides code lens.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) code_lens_provider: Option<CodeLensOptions>,
+
+    /// The server handles and provides custom nargo messages.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) nargo: Option<NargoCapability>,
+}
+
+#[derive(Debug, PartialEq, Clone, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct InitializeResult {
+    /// The capabilities the language server provides.
+    pub(crate) capabilities: ServerCapabilities,
+
+    /// Information about the server.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) server_info: Option<ServerInfo>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(try_from = "String", into = "String")]
+pub(crate) struct NargoTestId(CrateName, String);
```

Comment:
Why not use a struct for this?

Reply:
No reason. I tried a bunch of stuff with Serde and didn't like the behavior of anything except `TryFom<String>` so this can be defined with named struct fields.